src/devices/src/virtio/net/device.rs

```diff
@@ -257,84 +276,94 @@ impl Net {
         success
     }
 
-    // Copies a single frame from `self.rx_frame_buf` into the guest. Returns true
-    // if a buffer was used, and false if the frame must be deferred until a buffer
-    // is made available by the driver.
-    fn rx_single_frame(&mut self) -> bool {
+    // Copies a single frame from `self.rx_frame_buf` into the guest.
+    fn do_write_frame_to_guest(&mut self) -> std::result::Result<(), FrontendError> {
+        let mut result: std::result::Result<(), FrontendError> = Ok(());
         let mem = match self.device_state {
             DeviceState::Activated(ref mem) => mem,
             // This should never happen, it's been already validated in the event handler.
             DeviceState::Inactive => unreachable!(),
         };
-        let rx_queue = &mut self.queues[RX_INDEX];
-        let mut next_desc = rx_queue.pop(mem);
-        if next_desc.is_none() {
-            METRICS.net.no_rx_avail_buffer.inc();
-            return false;
-        }
 
-        // We just checked that the head descriptor exists.
-        let head_index = next_desc.as_ref().unwrap().index;
-        let mut write_count = 0;
-
-        // Copy from frame into buffer, which may span multiple descriptors.
-        loop {
-            match next_desc {
-                Some(desc) => {
-                    if !desc.is_write_only() {
-                        break;
-                    }
-
-                    let limit = cmp::min(write_count + desc.len as usize, self.rx_bytes_read);
-                    let source_slice = &self.rx_frame_buf[write_count..limit];
-                    let write_result = mem.write_slice(source_slice, desc.addr);
+        let queue = &mut self.queues[RX_INDEX];
+        let head_descriptor = queue.pop(mem).ok_or_else(|| {
+            METRICS.net.no_rx_avail_buffer.inc();
+            FrontendError::EmptyQueue
+        })?;
+        let head_index = head_descriptor.index;
+
+        let mut frame_slice = &self.rx_frame_buf[..self.rx_bytes_read];
+        let frame_len = frame_slice.len();
+        let mut maybe_next_descriptor = Some(head_descriptor);
+        while let Some(descriptor) = &maybe_next_descriptor {
```

Comment:
```
5.1.6.4 Processing of Incoming Packets

When a packet is copied into a buffer in the receiveq, the optimal path is to disable further used buffer notifications for the receiveq and process packets until no more are found, then re-enable them.

Processing incoming packets involves:

    num_buffers indicates how many descriptors this packet is spread over (including this one): this will always be 1 if VIRTIO_NET_F_MRG_RXBUF was not negotiated. This allows receipt of large packets without having to allocate large buffers: a packet that does not fit in a single buffer can flow over to the next buffer, and so on. In this case, there will be at least num_buffers used buffers in the virtqueue, and the device chains them together to form a single packet in a way similar to how it would store it in a single buffer spread over multiple descriptors. The other buffers will not begin with a struct virtio_net_hdr.
    If num_buffers is one, then the entire packet will be contained within this buffer, immediately following the struct virtio_net_hdr. 
```

We do not negotiate for `VIRTIO_NET_F_MRG_RXBUF` and this means that the packets will always fit a buffer, which is represented by a single descriptor. I think this means that we do not need to process descriptor chains after all and can directly use a single descriptor to hold incoming packets. @serban300 , have you verified if the packets are written to a descriptor chain ( >1 available descriptors) or to a single one?

Reply:
I tried this and I'm getting descriptor chains with multiple buffer of size 12 or 4080. I'm not sure why. What you're saying might be part of a new version of the virtio standard. We're using 0.95 as far as I know. Anyway, I would keep the logic for processing descriptor chains since it's already implemented and we might need it in the future.